File: modelo_unificado_naive_bayes.R
# =====================================================================
# 🚀 ALGORITMO: NAIVE BAYES (EQUIVALENTE EM R)
# =====================================================================

# install.packages("e1071")
# install.packages("caret")
library(e1071)
library(caret)

# Função auxiliar para exibir as métricas
avaliar_modelo <- function(matriz_confusao, tipo_base) {
  dados_metricas <- confusionMatrix(matriz_confusao)
  acuracia <- dados_metricas$overall['Accuracy'] * 100
  
  if (tipo_base == "prof") {
    precisao <- dados_metricas$byClass['Precision'] * 100
    revocacao <- dados_metricas$byClass['Sensitivity'] * 100
    f1 <- dados_metricas$byClass['F1'] * 100
  } else {
    pesos <- colSums(matriz_confusao) / sum(matriz_confusao)
    precisao <- sum(dados_metricas$byClass[, 'Precision'] * pesos, na.rm = TRUE) * 100
    revocacao <- sum(dados_metricas$byClass[, 'Sensitivity'] * pesos, na.rm = TRUE) * 100
    f1 <- sum(dados_metricas$byClass[, 'F1'] * pesos, na.rm = TRUE) * 100
  }
  
  cat(sprintf("  Acurácia  : %.2f%%\n", acuracia))
  cat(sprintf("  Precisão  : %.2f%%\n", precisao))
  cat(sprintf("  Revocação : %.2f%%\n", revocacao))
  cat(sprintf("  F1-Score  : %.2f%%\n", f1))
  cat("  Matriz de Confusão:\n  (Linha: Real, Coluna: Previsto)\n")
  print(t(matriz_confusao))
  cat(strrep("-", 40), "\n")
}

print("=== RUNNING NAIVE BAYES PIPELINE ===")

# --- BASE 1: PROFESSOR ---
base_prof <- read.csv("credit_data.csv")
base_prof$age[base_prof$age < 0] <- 40.92
base_prof$loan[is.na(base_prof$loan)] <- mean(base_prof$loan, na.rm = TRUE)

X_prof_scaled <- scale(base_prof[, 2:4])
base_prof_final <- data.frame(X_prof_scaled, default = as.factor(base_prof$default))

set.seed(0)
amostra_p <- sample(1:nrow(base_prof_final), size = round(0.75 * nrow(base_prof_final)))
treino_p <- base_prof_final[amostra_p, ]
teste_p  <- base_prof_final[-amostra_p, ]

nb_prof <- naiveBayes(default ~ ., data = treino_p)
prev_nb_p <- predict(nb_prof, newdata = teste_p)

cat("\n📊 === RESULTADOS: BASE DO PROFESSOR ===\n")
avaliar_modelo(table(teste_p$default, prev_nb_p), "prof")

# --- BASE 2: KAGGLE ---
base_kaggle <- read.csv("Credit Score Classification Dataset.csv")
dummies <- dummyVars(" ~ .", data = base_kaggle[, names(base_kaggle) != "Credit.Score"])
X_kaggle_dummies <- predict(dummies, newdata = base_kaggle)
X_kaggle_scaled <- scale(X_kaggle_dummies)
base_kaggle_final <- data.frame(X_kaggle_scaled, Credit.Score = as.factor(base_kaggle$Credit.Score))

set.seed(1)
amostra_k <- sample(1:nrow(base_kaggle_final), size = round(0.75 * nrow(base_kaggle_final)))
treino_k <- base_kaggle_final[amostra_k, ]
teste_k  <- base_kaggle_final[-amostra_k, ]

nb_kagg <- naiveBayes(Credit.Score ~ ., data = treino_k)
prev_nb_k <- predict(nb_kagg, newdata = teste_k)

cat("\n📊 === RESULTADOS: BASE DO KAGGLE ===\n")
avaliar_modelo(table(teste_k$Credit.Score, prev_nb_k), "kaggle")

File: modelo_unificado_naive_bayes.py
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score


print(" INICIANDO PIPELINE DE INTELIGÊNCIA ARTIFICIAL ")

# PARTE 1: PROCESSAMENTO DA BASE DE DADOS credit_data.csv
print("\n[BASE 1] Carregando dados do professor (credit_data.csv)...")
base_prof = pd.read_csv('credit_data.csv')

# faxina dos dados credit_data.csv (tratando idades negativas e nulos de forma direta)
base_prof.loc[base_prof['age'] < 0, 'age'] = 40.92

# TRUQUE DEFINITIVO: preenche os nulos direto no array numérico com o Pandas
base_prof = base_prof.fillna(base_prof.mean(numeric_only=True))
#X_prof = pd.DataFrame(X_prof).fillna(base_prof['loan'].mean()).values

# separando X e y (usando as colunas numéricas de renda, idade e empréstimo)
X_prof = base_prof.iloc[:, 1:4].values
y_prof = base_prof.iloc[:, 4].values

# Verificando se ainda existem nulos na base do professor
#print("Valores nulos por coluna:\n", base_prof.isnull().sum())

# escalonamento dos dados
scaler = StandardScaler()
X_prof = scaler.fit_transform(X_prof)
# divisão treino/teste
X_tr_p, X_te_p, y_tr_p, y_te_p = train_test_split(X_prof, y_prof, test_size=0.25, random_state=0)

# treinamento do algoritmo
gnb_prof = GaussianNB()
gnb_prof.fit(X_tr_p, y_tr_p)
prev_prof = gnb_prof.predict(X_te_p)

# exibindo resultados da base 1
print("\n📊 === RESULTADOS: BASE DO PROFESSOR ===")
print(f"Acurácia : {accuracy_score(y_te_p, prev_prof) * 100:.2f}%")
print(f"Precisão : {precision_score(y_te_p, prev_prof) * 100:.2f}%")
print(f"Revocação : {recall_score(y_te_p, prev_prof) * 100:.2f}%")
print(f"F1-Score : {f1_score(y_te_p, prev_prof) * 100:.2f}%")
print("Matriz de Confusão:")
print(confusion_matrix(y_te_p, prev_prof))


print("\n--------------------------------------------------")
print("[BASE 2] Carregando dados do Kaggle (Credit Score Classification Dataset.csv)...")
base_kaggle = pd.read_csv('Credit Score Classification Dataset.csv')

# preparação dos dados do Kaggle (Usando o get_dummies para os textos)
y_kaggle = base_kaggle['Credit Score'].values
X_tab_kaggle = pd.get_dummies(base_kaggle.drop('Credit Score', axis=1))
X_kaggle = X_tab_kaggle.values

# escalonamento dos dados
X_kaggle = scaler.fit_transform(X_kaggle)

# divisão treino/teste
X_tr_k, X_te_k, y_tr_k, y_te_k = train_test_split(X_kaggle, y_kaggle, test_size=0.25, random_state=1)

# treinamento do algoritmo
gnb_kaggle = GaussianNB()
gnb_kaggle.fit(X_tr_k, y_tr_k)
prev_kaggle = gnb_kaggle.predict(X_te_k)

# exibindo resultados da base 2
print("\n📊 === RESULTADOS: BASE DO KAGGLE ===")
print(f"Acurácia : {accuracy_score(y_te_k, prev_kaggle) * 100:.2f}%")
print(f"Precisão : {precision_score(y_te_k, prev_kaggle, average='weighted', zero_division=0) * 100:.2f}%")
print(f"Revocação : {recall_score(y_te_k, prev_kaggle, average='weighted', zero_division=0) * 100:.2f}%") 
print(f"F1-Score : {f1_score(y_te_k, prev_kaggle, average='weighted', zero_division=0) * 100:.2f}%")
print("Matriz de Confusão:")
print(confusion_matrix(y_te_k, prev_kaggle))
